### ISSUE TITLE
TypeScript cannot find declaration files with module nodenext

### ISSUE THREAD

**Issue by cseas** [razorpay/blade#960]
## Problem

Blade ships with `d.ts` declaration files. However, since the bundled `package.json` doesn't contain a `"types"` field, the consumer project is unable to find these declaration files and as a result ignores the types for components that Blade ships.

VSCode can't find the declaration files either as shown by the below warning:

![Screenshot 2023-01-17 at 1 10 50 PM](https://user-images.githubusercontent.com/29686866/212837905-bf70df19-9bab-4109-805a-ee28d96ca173.png)

This means that IntelliSense in the editor doesn't work for types/props. Also, passing wrong props to a component works. For example, the below component compiles without errors.

```tsx
<Button hello="world">Pay Now</Button>
```

## Solution

The `package.json` should contain a ["types"][1] field that points to the declaration file so the bundler and IDE can find it.

Refer ESM guide for types in `exports` field: https://www.typescriptlang.org/docs/handbook/esm-node.html

Note that the *fallback* types field in the above link is required even for ESM bundles right now because proper support for `exports` field is a part of TypeScript 5 milestone: https://github.com/microsoft/TypeScript/issues/50794#event-8030311249

[1]: https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html#including-declarations-in-your-npm-package

**Comment by divyanshu013** [comment]
Interesting, I wonder how's it working for other projects right now 🤔

**Comment by kamleshchandnani** [comment]
from the docs link you mentioned

>Also note that if your main declaration file is named `index.d.ts` and lives at the root of the package (next to index.js) you do not need to mark the types property, though it is advisable to do so.


Can you link a possible codesandbox repro? 

so far we haven't faced this issue ourselves while using it in multiple scaffolded projects and consumer projects too and none of our consumers have also raised it.

**Comment by cseas** [comment]
Here's a reproduction:
- Clone https://github.com/cseas/repro/
- `git checkout repro-959`
- `npm i`
Opening `src/App.tsx` in VSCode should show the warning.
- `npm run build` will run fine regardless what props you pass to Button.

I've tried installing other projects that ship with types and types for those get detected correctly. VSCode only shows this problem with Blade.

**Comment by kamleshchandnani** [comment]
can you check once and post what version of TS server is set on your vscode?

**Comment by cseas** [comment]
v4.7.3

<img width="329" alt="Screenshot 2023-01-18 at 1 36 49 PM" src="https://user-images.githubusercontent.com/29686866/213117008-31d85541-cf4a-4a23-9ccd-dd7646e7a53b.png">

**Comment by kamleshchandnani** [comment]
can you switch to 4.9? because we had added module-specific resolution for `.native` and `.web` extensions as that feature was released in 4.8 so that could be one issue

**Comment by cseas** [comment]
I did try v4.9.4, not working 😅 

![Screenshot 2023-01-18 at 4 46 33 PM](https://user-images.githubusercontent.com/29686866/213158643-e0615706-bb72-48f9-945f-60f3fa3d9ee7.png)

Tried reloading VSCode after switching TypeScript version too, issue persists.

**Comment by anuraghazra** [comment]
@cseas I noticed you didn't specify `moduleResolution` in tsconfig, when you don't specify it and set `module=NodeNext` it defaults to `Classic`. 

If you set `moduleResolution: 'Node'` it should work as expected, the `Classic` option is set as default for compatibility reasons with older versions of TS.

**Comment by cseas** [comment]
> `moduleResolution` in tsconfig, when you don't specify it and set `module=NodeNext` it defaults to `Classic`.

No, moduleResolution takes the [same value as module][1] when module is set to `NodeNext`. So moduleResolution in the linked repro is `NodeNext` which is intended.

<img width="980" alt="Screenshot 2023-01-20 at 1 25 25 PM" src="https://user-images.githubusercontent.com/29686866/213645183-e9860150-2bcb-41a9-9f83-24c8dcc6e1ca.png">

Changing the `module` or `moduleResolution` to an option which transpiles to CommonJs would indeed resolve the issue. But that changes our bundle output to CommonJS instead of ESM, which is undesired.

More on this: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c#how-can-i-move-my-commonjs-project-to-esm

Also, as [pointed out in this comment][2], TypeScript finds declaration files correctly for other UI libraries. So the problem isn't the setup/config here, the issue is that Blade is not supporting a setup that wants to bundle as ESM.

[1]: https://www.typescriptlang.org/tsconfig#moduleResolution
[2]: https://github.com/razorpay/blade/issues/960#issuecomment-1385533979

**Comment by anuraghazra** [comment]
Okay then in that case we can check if types field resolves the issue, but we need to see if it can work with moduleSuffixes since it needs to work seamlessly across RN and web.

**Comment by cseas** [comment]
Yes, the RN support is what I'm not sure about. The solution I suggested in the issue description will solve it for web projects but that could break the types for RN.

**Comment by kamleshchandnani** [comment]
@cseas from the codesandbox repo you sent above I see things are working fine 🤔 
Played around with different values for `module` and it works in all cases even tried without changing anything on the link you have attached

<img width="708" alt="Screenshot 2023-01-25 at 11 26 14 AM" src="https://user-images.githubusercontent.com/11384858/214491065-da51036f-ca33-4799-9ef8-793dab85a221.png">

**Comment by cseas** [comment]
My bad. I was playing around with the tsconfig values to debug another issue. Didn't realise Codesandbox saved it. Try now, the issue should come up.

Here's the tsconfig that reproduces the issue:

```json
{
    "compilerOptions": {
        "jsx": "react-jsx",
        "module": "NodeNext"
    }
}
```

**Comment by kamleshchandnani** [comment]
Another question: to make this work as per your [doc link](https://www.typescriptlang.org/docs/handbook/esm-node.html#type-in-packagejson-and-new-extensions) `type: module` should be set in target of blade but we haven't set that so right now with your fix anyways things might break while bundling if the consumer project sets `type: module` in their package.json because now its misleading since our exports field is only pointing to `.js` version even though the output in the `.js` files is esm but its not following the file extension. so this might work for this particular use case but I'm not sure if this will work flawlessly for projects which set `type: module` in their package.json 🤔

**Comment by cseas** [comment]
That's correct, ESM version of libs should either ship with the `.mjs` extension or with `type: module` for node to recognise it as ESM. Not sure how it's working right now, need to investigate more on it. 

My assumption is that there's no issues because the `exports` field in Blade's `package.json` directly points to the single flat file that contains everything a user can import from Blade. The module resolution of node will probably only start throwing errors on relative imports once Blade starts shipping the original tree instead of a single file (required to solve #959).

The GitHub repro I linked above has `type: module`. I manually added the changes from my linked PR to Blade under `node_modules` and tested. Just added `type: module` to the linked Codesandbox and that works too. So bundle looks fine. The only problem is TypeScript declarations not being found.

Related discussion: https://github.com/microsoft/TypeScript/issues/18442#issuecomment-1369110873
